src/ServiceInsight/MessageFlow/MessageFlowViewModel.cs

```diff
@@ -259,18 +260,39 @@ void LinkConversationNodes(IEnumerable<MessageNode> relatedMessagesTask)
         {
             foreach (var msg in relatedMessagesTask)
             {
-                if (msg.Message.RelatedToMessageId == null &&
-                    msg.Message.RelatedToMessageId != msg.Message.MessageId)
+                if (string.IsNullOrEmpty(msg.Message.RelatedToMessageId) && msg.Message.RelatedToMessageId != msg.Message.MessageId)
                 {
                     continue;
                 }
 
-                // [CM] I don't know how it's happening, but a user reported an
-                // error where multiple results were returned from this query.
                 var parentMessages = nodeMap.Values.Where(m =>
-                    m.Message != null && m.Message.ReceivingEndpoint != null && m.Message.SendingEndpoint != null &&
-                    m.Message.MessageId == msg.Message.RelatedToMessageId &&
-                    m.Message.ReceivingEndpoint.Name == msg.Message.SendingEndpoint.Name);
+                    m.Message != null && m.Message.ReceivingEndpoint != null && m.Message.SendingEndpoint != null
+                    && m.Message.MessageId == msg.Message.RelatedToMessageId
+                    && m.Message.ReceivingEndpoint.Name == msg.Message.SendingEndpoint.Name // Needed with publishes as identical events can be consumed by multiple endpoints
+                    );
+
+                // Fallback, get "parent" when originating message is not an event (publish)
+                if (!parentMessages.Any())
+                {
+                    parentMessages = nodeMap.Values.Where(m =>
+                        m.Message != null && m.Message.ReceivingEndpoint != null && m.Message.SendingEndpoint != null &&
+                        m.Message.MessageId == msg.Message.RelatedToMessageId && m.Message.MessageIntent != MessageIntent.Publish
+                        );
+
+                    if (parentMessages.Any())
+                    {
+                        LogTo.Warning("Fall back to match only on RelatedToMessageId for message {0} matched but link could be invalid.", msg.Message.MessageId);
+                    }
+                }
+
+                if (!parentMessages.Any())
+                {
+                    LogTo.Error("No parent could be resolved for message {0} which has RelatedToMessageId set. This can happen if the parent has been purged due to retention expiration, an ServiceControl node to be unavailable, or because the parent message not been stored (yet).", msg.Message.MessageId);
+                }
+                else if (parentMessages.Count() > 1)
+                {
+                    LogTo.Error("Multiple parents matched for {0} possibly due to more-than-once processing, linking to all as unknown which processing attemps generated the message.", msg.Message.MessageId);
```

Comment:
Where are errors logged to? Should we tone this down a bit? What would a user do with an error and why should they care?

Reply:
ditto https://github.com/Particular/ServiceInsight/pull/1171#discussion_r826716375

Reply:
@danielmarbach Logs are outputted in SI in the logs tab: https://docs.particular.net/serviceinsight/logging

At least the error/warning indicates that the data visualized is correct or in this case an explanation for the strange multiple connections. Yes, it's not ideal but it's better than nothing.